### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by maximburyak**

### CODE REVIEW

File: Jint/Native/Array/ArrayInstance.cs [hunk 1]
```diff
@@ -39,6 +40,31 @@ public ArrayInstance(Engine engine, uint capacity = 0) : base(engine)
             }
         }
 
+        public ArrayInstance(Engine engine, PropertyDescriptor[] items) : base(engine)
+        {
+            _engine = engine;
+            if (items.Length < MaxDenseArrayLength)
+            {
+                _dense = (items != null && items.Length !=0 ) ? items : System.Array.Empty<PropertyDescriptor>();
```
Comment (ayende) [inline Jint/Native/Array/ArrayInstance.cs]: items cannot be null here, if it is, we already thrown NRE

File: Jint/Native/Array/ArrayInstance.cs [hunk 2]
```diff
@@ -20,7 +21,7 @@ public class ArrayInstance : ObjectInstance, IEnumerable<JsValue>
 
         private PropertyDescriptor _length;
 
-        private const int MaxDenseArrayLength = 1024 * 10;
+        public const int MaxDenseArrayLength = 1024 * 10;
```
Comment (ayende) [inline Jint/Native/Array/ArrayInstance.cs]: Why is this public?

**Review by ayende — CHANGES_REQUESTED** [review]

File: Jint/Native/Array/ArrayInstance.cs (hunk 2 above)
Comment (maximburyak) [reply Jint/Native/Array/ArrayInstance.cs]: We need to know whether we want to send and array or a dictionary to the ctor.

Comment (lahma) [reply Jint/Native/Array/ArrayInstance.cs]: Doe Raven ever actually need to jump from dense to sparse? Dictionary is mostly for JS madness where you can have indexes 0 and 1000000 without nothing in between (growth algorithm tries to detect over 50 index jumps and goes to sparse then..

Comment (ayende) [reply Jint/Native/Array/ArrayInstance.cs]: what is the source here? We are fine with having a dense array that is 50,000 items long

Comment (maximburyak) [reply Jint/Native/Array/ArrayInstance.cs]: First, I sincerely hope that there will not be documents with 10K long arrays :) About the dictionary, I don't think it's only about Raven.. The criteria in ArrayInstance whether to use a dictionary or an array is the collection size, which means that once your "capacity" is bigger then 10k, we are going to pay for Dictionary.Insert, which is _very_ expensive... On the other side, I understand the scenario with non-sequential arrays. I'll change the code to support array only or dictionary only ctors, while the question about how to differ between long sequential arrays and non-sequential arrays will stay open, and should be addressed in the "original" ctor

**Pull request merged**